docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+- It lacks buffering so it cannot be used reliably for input and output without flow control.
+- It pulls in the C library stdio system because it uses `Stream`.
+
+**UARTSerial**
+
+`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 
+
+It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it loses the ability to call printf directly.
+
+**RawSerial**
+
+`RawSerial` provides unbuffered I/O. It has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port.
+
+### Requirements and assumptions
+
+None
+
+# System architecture and high-level design
+
+Applications will have a choice between two public serial classes: 
+- `BufferedSerial` for a normal application
+- `UnbufferedSerial` for an interrupt driven application or one that needs to have more control
+
+### System architecture and component interaction
+
+The below diagram shows the inheritance hierarchy for the serial classes.
+
+![Serial classes](./diagrams/serial-classes.png)
+
+`MinimalSerial` is an abstract class that extracts basics I/O functionality from `SerialBase` so it can be re-used for a minimal serial console in the retarget code.
+
+# Detailed design
+
+### Detailed design : BufferedSerial
+
+`BufferedSerial` is `UARTSerial` renamed to convey the original purpose of the class. In addition, the new class will allow configurability of the transmit and receive buffer size so that each instance of the class can configure its buffer size according to its intended usage, for example a modem may need a large input buffer and a console may not need any input buffer. 
+
+The below diagram shows the detailed inheritance hierarchy for `BufferedSerial`.
+
+![BufferedSerial class](./diagrams/buffered-serial.png)
+
+**API description**
+
+This is the new constructor for `BufferedSerial` to pass in the serial buffers size.
+
+```C
+    /** Create a BufferedSerial port, connected to the specified transmit and receive pins, with a particular baud rate.
+     *  @param tx Transmit pin
+     *  @param rx Receive pin
+     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
+     *  @param rx_buffer_size the size of the receive buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE)
+     *  @param tx_buffer_size the size of the transmit buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE)
+     */
+    BufferedSerial(PinName tx, 
+                   PinName rx, 
+                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE, 
+                   uint32_t rx_buffer_size = MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE,
```

Comment:
As @bulislaw has reminded me a couple of times, we mustn't force dynamic allocation. This one would need to dynamically allocate the buffers, but we can also have constructors that pass in a buffer pointer:

    BufferedSerial(PinName tx, 
                   PinName rx, 
                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE, 
                   void *rx_buffer,
                   size_t rx_buffer_size,
                   void *tx_buffer,
                   size_t tx_buffer_size);

(Note there's no reason not to use size_t).

Also note that the recent pinmap stuff will multiply up the number of constructors...

Reply:
Good point.